internal/registry.go

```diff
@@ -0,0 +1,287 @@
+// Copyright (c) 2017-2020 Uber Technologies Inc.
+// Portions of the Software are attributed to Copyright (c) 2020 Temporal Technologies Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package internal
+
+import (
+	"fmt"
+	"reflect"
+	"strings"
+	"sync"
+)
+
+var once sync.Once
+
+// Singleton to hold the host registration details.
+var globalRegistry *registry
+
+func newRegistry() *registry {
+	return &registry{
+		workflowFuncMap:  make(map[string]interface{}),
+		workflowAliasMap: make(map[string]string),
+		activityFuncMap:  make(map[string]activity),
+		activityAliasMap: make(map[string]string),
+		next:             getGlobalRegistry(),
+	}
+}
+
+func getGlobalRegistry() *registry {
+	once.Do(func() {
+		globalRegistry = &registry{
+			workflowFuncMap:  make(map[string]interface{}),
+			workflowAliasMap: make(map[string]string),
+			activityFuncMap:  make(map[string]activity),
+			activityAliasMap: make(map[string]string),
+		}
+	})
+	return globalRegistry
+}
+
+type registry struct {
```

Comment:
nit: move struct to top

Reply:
I leave them here because I'll make them public methods if I move the file into its own package.